Pallet executor registry (part 2)

liuchengxu opened the pull request:
This PR primarily achieves these goals:

- Implement `update_public_key` in pallet-executor-registry.
- Rotate the executor authorities for the bundle production on each epoch.
- Implement the new stake-weight-based bundle election for producing a bundle on the executor end.
- Verify the new bundle election on the farmer end.

This PR serves as a prototype instead of the final product, there are a few TODOs left. The general workflow should be complete, there are some details I still need to check with the spec to ensure they are aligned later. I ran a few executor tests locally, the new logic seems to work fine, but more testing is needed for sure. Walking the commits one by one should make the most sense.

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](../CONTRIBUTING.md)

liuchengxu committed:
Impl `update_public_key`

liuchengxu committed:
Rotate the executor authorities on each new epoch

liuchengxu committed:
Track slot probability in pallet-executor-registry

The slot probability of each domain will be required to calculate the
bundle election threshold later. Track it in the state so that it can be
stored in the storage witness needed by the farmer to verify the bundle.

liuchengxu committed:
Initialize Authorities/TotalStakeWeight/SlotProbability at genesis

liuchengxu committed:
Integrate pallet-executor-registry into secondary runtime

liuchengxu committed:
Small refactoring to prepare the new bundle election

liuchengxu committed:
Update the signature of `solve_bundle_election_challenge()`

`global_challenge` will be used as the gloabl randomness to derive the
local domain randomness for the bundle election.

liuchengxu committed:
Impl the new bundle election logic for producing a bundle

liuchengxu committed:
Introduce ProofOfElection in the bundle

The generation of ProofOfElection is done in this commit, the
verification will be in the next commit.

liuchengxu committed:
Validate ProofOfElection in the bundle on primary chain

Basically, we need to check:

1. vrf signature is correct.
2. bundle election solution satisfies the threshold.

liuchengxu committed:
Make tests compile

nazar-pc commented on cumulus/pallets/executor-registry/src/lib.rs:
```diff
@@ -423,8 +426,18 @@ mod pallet {
+        pub fn update_public_key(origin: OriginFor<T>, next_key: ExecutorId) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            ensure!(
+                KeyOwner::<T>::get(&next_key).is_none(),
+                Error::<T>::DuplicatedKey
+            );
```
Why do we care about this at all?

liuchengxu replied:
It's necessary so that the executor is accountable for the mishavor if something like the heartbeat will be possibly introduced in the future.

nazar-pc commented on cumulus/pallets/executor-registry/src/lib.rs:
```diff
@@ -423,8 +426,18 @@ mod pallet {
+                Error::<T>::DuplicatedKey
+            );
+
+            NextKey::<T>::insert(&who, &next_key);
+            KeyOwner::<T>::insert(&next_key, &who);
```
These are inserted, but there is no code added that ever removes those values. Also nothing is inserted into `KeyOwner` during registration either.

liuchengxu replied:
Good catch!

nazar-pc commented on cumulus/pallets/executor-registry/src/lib.rs:
```diff
@@ -621,9 +621,48 @@ mod pallet {
             if (block_number % T::EpochDuration::get()).is_zero() {
-                // TODO: Enact the epoch change.
+                let mut total_stake_weight = 0;
+                let authorities = Executors::<T>::iter()
```
This bothers me a lot. Are you sure we'll not have problems ever iterating over all executors?

I see that you clone all executors into `Authorities`.

I feel like we need to index them instead and have a mapping to those indexes, such that when we do the actual transition, we can deliberately query just the executors we want and not iterate over all of them. Can be a TODO for now.

liuchengxu replied:
It should be fine for 100k executors without a complex election logic, that's the number we're targeting in the mid-term feature if I recall right, I haven't done the benchmarking how many weights are actually required though. Since we don't expect there are too many inactive executors actually, not sure how much performance improvements we can achieve. Let's left a TODO here and I believe we will figure it out at least when we have proper weights.

nazar-pc commented on cumulus/pallets/executor-registry/src/lib.rs:
```diff
@@ -621,9 +621,48 @@ mod pallet {
+                                    let executor_config = maybe_executor_config
+                                        .as_mut()
+                                        .expect("Executor config must exist; qed");
+                                    executor_config.public_key = new_key.clone();
```
Just a though: since you already have `executor_config`, you could do something like this:
```suggestion
                                    executor_config.public_key = new_key.clone();
                                    maybe_executor_config.replace(executor_config.clone());
```

liuchengxu replied:
Hmm, that needs to clone the whole executor config though, even it's likely not significant.

nazar-pc commented on crates/sp-executor/src/lib.rs:
```diff
@@ -59,6 +59,9 @@ impl sp_runtime::BoundToRuntimeAppPublic for ExecutorKey {
 }
 
+/// Stake weight in the bundle election.
+pub type StakeWeight = u128;
```
What about using balance generic in the pallet instead of explicit value that balance need to be able to fit to anyway? Or make it generic type in pallet that supports `From<Balance>`?

liuchengxu replied:
My initial idea was to not have the generic if possible, but now it seems to be helpful to use the generic type considering the conversions in the runtime. Using a new generic type that supports `From<Balance>` than the balance generic directly is semantically clearer.

liuchengxu replied:
@nazar-pc Except for this comment, the others can be reviewed again. I need to do some experiments to see if it's worthy of using the generic here, my intuition is still uncertain about it.

liuchengxu replied:
Please see if https://github.com/subspace/subspace/pull/861/commits/9d930e902f9c7b0d2c1f0c95b49dff1dbfc655cb is good for you, @nazar-pc @vedhavyas

nazar-pc commented on cumulus/pallets/executor-registry/src/lib.rs:
```diff
@@ -658,6 +697,17 @@ mod pallet {
+    /// Current epoch executor authorities.
+    #[pallet::storage]
+    #[pallet::getter(fn authorities)]
+    pub(super) type Authorities<T: Config> =
```
`SelectedExecutors`? `Authorities` reminds me of PoA, these however do not have any real authority, they are just eligible to do execution.

liuchengxu replied:
That's kind of subjective. I actually had a small discussion with Jeremiah on this point before, and we think it's okay to use the term authority here as it's not as permissionless as the farmer.

nazar-pc commented on cumulus/runtime/src/runtime.rs:
```diff
@@ -243,6 +243,31 @@ impl cirrus_pallet_executive::Config for Runtime {
+    pub const EpochDuration: BlockNumber = 600;
+    pub const MaxWithdrawals: u32 = 1;
+    // One day.
+    pub const WithdrawalDuration: BlockNumber = 14400;
```
I typically write components explicitly rather than the resulting number:
```suggestion
    // One hour in blocks.
    pub const EpochDuration: BlockNumber = 3600 / 6;
    pub const MaxWithdrawals: u32 = 1;
    // One day in blocks.
    pub const WithdrawalDuration: BlockNumber = 3600 * 24 / 6;
```

nazar-pc commented on cumulus/client/cirrus-executor/src/bundle_producer.rs:
```diff
@@ -95,6 +95,11 @@ where
     > {
+        let ExecutorSlotInfo {
+            slot,
+            global_challenge,
```
Should probably be called `slot_randomness` here as well then.

nazar-pc commented on crates/sp-executor/src/lib.rs:
```diff
@@ -102,6 +113,62 @@ impl<Hash: Encode> BundleHeader<Hash> {
+    let mut data = domain_id.to_le_bytes().to_vec();
+    data.extend_from_slice(vrf_output);
+
+    let local_domain_randomness = blake2b_256_hash(&data);
```
There is a function that hashes a list of values, you don't need to allocate a vector like that just to hash it

nazar-pc commented on crates/sp-executor/src/lib.rs:
```diff
@@ -102,6 +113,62 @@ impl<Hash: Encode> BundleHeader<Hash> {
+    slot_probability: (u64, u64),
+) -> u128 {
+    stake_weight / total_stake_weight * u128::MAX * u128::from(slot_probability.0)
+        / u128::from(slot_probability.1)
```
This will always be zero assuming `stake_weight` is never as big as `total_stake_weight`. Also `* u128::MAX` will cause overflow unless `stake_weight / total_stake_weight` is `0` or `1`.

This is problematic, I think you need to rewrite it correctly, possibly like this:
```suggestion
    u128::MAX / u128::from(slot_probability.1) * u128::from(slot_probability.0) / total_stake_weight * stake_weight
```

And provide a comment what it is in more readable form, I did that in `pallet-subspace` already.

liuchengxu replied:
Thanks for catching this! It wasn't caught earlier due to the special case `stake_weight / total_stake_weight` is `1` in the few tests I ran before, haven't done more tests which will be worked soon.

nazar-pc commented on cumulus/client/cirrus-executor/src/bundle_producer.rs:
```diff
@@ -219,6 +220,57 @@ where
+                    slot_probability,
+                );
+
+                if u128::from(election_solution) <= threshold {
```
I don't think I follow why do you have `threshold` as `u64` if you compare it against potentially much larger value?

liuchengxu replied:
Using `u64` for the solution is defined in the spec, the threshold is based on the stake weight which is from the u128 balance. That's something I'd like to confirm with Jeremiah later, I left a TODO on unifying the type between threhold and election solution, will have a result before the next integrations sync.

liuchengxu replied:
Unified both `threshold` and `solution` to u128.

nazar-pc commented on crates/sp-executor/src/lib.rs:
```diff
@@ -102,6 +113,62 @@ impl<Hash: Encode> BundleHeader<Hash> {
 }
 
+// TODO: unify the type between the solution and threshold better.
+pub fn derive_bundle_election_solution(domain_id: DomainId, vrf_output: &[u8]) -> u64 {
```
This isn't the right way of doing it, you need something like this (note the APIs used):
https://github.com/subspace/subspace/blob/e18b357254d4e4c766b90f873ede55c318c22502/crates/subspace-verification/src/lib.rs#L291-L306

liuchengxu replied:
That's something I'm certain about and would like to figure out. I recall you ran into a problem that the vrf signature wasn't used properly before, so I paid some attention to it, the verification seemed to work fine. I wanted to have a test case to ensure the vrf proof usage is fine, but I haven't figured out such a test case, do you already have it somethere in the farmer code?

nazar-pc replied:
Verification will work properly, but it may not be secure. There was a discussion about this somewhere already, TL;DR is that you should use the code I referenced above to do it correctly, this is also what BABE uses for similar purposes.

liuchengxu replied:
I'll update it for sure, just wanted to make sure the usage is proper in the code, that's why I asked if there is a test we can add here to ensure that, otherwise it's sort of unwritten knowledge known in a small group of people that using the vrf proof needs extra care.

nazar-pc commented on crates/sp-executor/src/lib.rs:
```diff
@@ -169,6 +202,38 @@ pub struct BundleElectionParams {
+    pub vrf_output: Vec<u8>,
+    /// VRF proof.
+    pub vrf_proof: Vec<u8>,
+    /// VRF public key.
+    pub vrf_public_key: Vec<u8>,
```
Might be called `executor_public_key`

liuchengxu replied:
Actually, I has been planning renaming `ExecutorId` to `ExecutorPublicKey` so that it's consistent with the `FarmerPublicKey` as they are quite similar regarding the signing purpose. I'll probably rename it when I resolve the TODO on merging the `executor_id` and `vrf_public_key`.

nazar-pc commented on cumulus/client/cirrus-executor/src/bundle_producer.rs:
```diff
@@ -264,10 +270,44 @@ where
+
+                    // TODO: vrf_public_key and authority_id are essentially the same, merge them?
+                    let vrf_public_key =
+                        schnorrkel::PublicKey::from_bytes(&ByteArray::to_raw_vec(&authority_id))
```
Doesn't `authority_id.as_ref()` exist?

nazar-pc commented on cumulus/client/cirrus-executor/src/bundle_producer.rs:
```diff
@@ -264,10 +270,44 @@ where
-                    // TODO: Add ProofOfElection into bundle so that farmer can verify the bundle
-                    // election.
+                    let storage_keys = well_known_keys::bundle_election_storage_keys();
+                    let storage_proof = self.client.read_proof(
+                        &best_block_id,
+                        &mut storage_keys.iter().map(|s| s.as_slice()),
+                    )?;
```
Hm... looks like this "proof" actually contains all the data within it. Isn't it a problem that such "proof" is going to be huge? I was expecting that there would be proofs that certain values are within state root instead, like 3 Merkle Tree proofs and all the inputs to those, rather than the full maps/vectors.

liuchengxu replied:
Make sense, but not feasible this time as the entire authoritiy set is stored as a single storage value in the runtime, let's use a TODO here and think about it on the next iteration, will make it a relative priority on my list.

nazar-pc commented on crates/sp-executor/src/lib.rs:
```diff
@@ -202,6 +203,81 @@ pub struct BundleElectionParams {
+    state_root: &H256,
+) -> Result<BundleElectionParams, ReadBundleElectionParamsError> {
+    use well_known_keys::{AUTHORITIES, SLOT_PROBABILITY, TOTAL_STAKE_WEIGHT};
+    use ReadBundleElectionParamsError::{DecodeError, MissingValue, TrieError};
```
We generally don't use in-function imports unless they are platform-specific or otherwise conditionally compled. Also I don't think we used enum variant imports before either.

nazar-pc commented on crates/pallet-executor/src/lib.rs:
```diff
@@ -586,6 +596,58 @@ impl<T: Config> Pallet<T> {
+            slot_probability,
+        );
+
+        if u128::from(election_solution) > threshold {
```
In farmer code what I did is abstract this away into a single function that returns `true` if within threshold, seems to be safer that way.

nazar-pc reviewed (commented):
Some questions, possible bugs and nits. Nits are not critical to be resolved of course.

I'd really love to see all commits to compile though, it is strange to see storage item introduced in one commit, used in another and passed in in the third. It makes it a bit harder to review because you can forget that the other two commit had by the time you get to the end, so jumping back and forth is required to make sense of changes.

liuchengxu committed:
Apply review suggestions

vedhavyas commented on crates/sp-executor/src/lib.rs:
```diff
@@ -59,6 +67,13 @@ impl sp_runtime::BoundToRuntimeAppPublic for ExecutorKey {
+
+/// Unique identifier for each domain.
+// TODO: unify the DomainId usage across the codebase.
+pub type DomainId = u64;
```
can we move this in to primitives and take the associated type with config ? This makes it easier to reuse this across multiple pallets

liuchengxu replied:
Can be done in another PR and that's the plan of the TODO above this line. To be honest, I'm not yet sold that it's really necessary to use the associate type in the config than a simple unit type wrapper or even an alias to a primitive like `u64`. Usually, we can have simpler code if it's possible to avoid the generic type.

vedhavyas replied:
Ideal would to be to have all the primitives under primitves crates. Regarding the associate type, its just preference to keep type not very concrete on pallet side. Could be in a different PR for sure

liuchengxu replied:
Yeah, it will be defined in the primitives crate for sure when the refactoring is done.

It's typical to use the associate type for the runtime specific type so that it's configurable for the runtime builder, but if it's a low level concept across the client and runtime, and no need to be configurable, using the concrete type is probably a better choice (always nicer to have simpler code) IMHO.

liuchengxu reviewed (commented):
>I'd really love to see all commits to compile though, it is strange to see storage item introduced in one commit, used in another and passed in in the third. 

Hmm, is there any commit that does not compile? It's unexpected if it does not compile, I tried to craft this PR with the care that the updates are incremental and the code can compile except for the tests.

vedhavyas commented on crates/sp-executor/src/lib.rs:
```diff
@@ -99,6 +115,213 @@ impl<Hash: Encode> BundleHeader<Hash> {
+    }
+}
+
+pub mod well_known_keys {
```
Would prefer deriving the key using the storage itself so that we dont have to change these when the storage changes.

liuchengxu replied:
We can do it on the secondary chain side, but it's not possible on the primary chain side because the primary chain pallet `pallet-executor` also needs to know the storage key to read the corresponding state in the storage proof but there is no way to access the secondary pallet info in a typed way on the primary chain side.

nazar-pc replied:
Sometimes it is more efficient to have hardcoded keys, if possible I just add a test case to make sure we catch regression, but avoid runtime overhead. I believe I did that a few times already.

vedhavyas commented on crates/sp-executor/src/lib.rs:
```diff
@@ -99,6 +115,213 @@ impl<Hash: Encode> BundleHeader<Hash> {
+    /// VRF proof.
+    pub vrf_proof: Vec<u8>,
+    /// VRF public key.
+    pub vrf_public_key: Vec<u8>,
```
Can we use AccountId type from frame_system instead of Vec or is it not a valid assumption to make ?

liuchengxu replied:
It's actually not an AccountId but an `ExecutorId` https://github.com/subspace/subspace/blob/585c940f417d360092c76ddf0dd760d7ffd657b5/crates/sp-executor/src/lib.rs#L62.

vedhavyas replied:
yes yes, I meant to ask if we can make the type more concrete instead of vec

liuchengxu replied:
I see, same thought here :( I'll try to make it typed when doing the refactoring later.

vedhavyas reviewed (commented):
Makes sense overall

liuchengxu committed:
Change the bundle election type to u128

liuchengxu committed:
Introduce the associate type StakeWeight in pallet-executor-registry

So that no need to forcibly convert the Balance into the StakeWeight
within the runtime.

nazar-pc commented on crates/sp-executor/src/lib.rs:
```diff
@@ -132,13 +130,26 @@ pub fn derive_bundle_election_solution(domain_id: DomainId, vrf_output: &[u8]) -
+    //
+    // TODO: better to have more audits on this calculation.
+    u128::MAX / total_stake_weight * u128::from(slot_probability.0) / u128::from(slot_probability.1)
+        * stake_weight
```
The way I wrote it made sure overflow logically never happens. Here if `total_stake_weight` is `1` and `slot_probability.0` is `123`, you'll still get an overflow. So it is better to divide and then multiply by the same kind of value to prevent such edge cases: `/ stake * stake / probability * probability`.

nazar-pc commented on cumulus/pallets/executor-registry/src/lib.rs:
```diff
@@ -629,6 +736,34 @@ mod pallet {
+
+    /// A map tracking the owner of current and next key of each executor.
+    #[pallet::storage]
+    pub(super) type KeyOwner<T: Config> =
```
I think there are still 2 edge-cases left:
* not inserted for executors in genesis config
* not removed on de-registration

liuchengxu replied:
Fixed the genesis config and noted the removal of KeyOwner in the deregister(TODO).

nazar-pc reviewed (commented):
> Hmm, is there any commit that does not compile? It's unexpected if it does not compile, I tried to craft this PR with the care that the updates are incremental and the code can compile except for the tests.

I didn't try, but I believe the commit where you updated genesis config fields, but not chain spec doesn't compile. Also tests are code, they should all compile, ideally without warnings. If they don't, it significantly hurts ability to trace regressions later.

liuchengxu committed:
Fix the KeyOwner insertion in genesis config and adjust the election solution threshold calculation

liuchengxu reviewed (commented):
>I didn't try, but I believe the commit where you updated genesis config fields, but not chain spec doesn't compile.

That's not the case. It compiles as when the genesis config was changed, the pallet was not yet integrated into the runtime, which was integrated in the subsequent commit, so chain spec is fine.

>Also tests are code, they should all compile, ideally without warnings. If they don't, it significantly hurts ability to trace regressions later.

It's valuable to add this to the contributing guide.

nazar-pc reviewed (approved):
Overall makes sense when TODOs are accounted for. Stake type is not critical.

> That's not the case.

I take my words back then, sorry for misinterpreting it.

> It's valuable to add this to the contributing guide.

I don't see why we should have special cases for tests or any other code. If repo doesn't compile - it is bad and needs to be fixed before committing.

liuchengxu committed:
Merge branch 'main' into pallet-executor-registry-part-2
